Add partial support for the 'Range' header when serving static files

atheriel opened the pull request:
As proposed in #259. I haven't done extensive 'real-world' testing, but I've been able to use this to enable rewinding/skipping in HTML `<audio>` tags in a Shiny app.

This PR includes

* Parsing and validation of the [most common forms](https://tools.ietf.org/html/rfc7233#section-3.1) of the `Range` header, namely `bytes=0-1000` and `bytes=1001-`, for serving slices of files and enabling resumable downloads, etc.

* Support for returning partial files matching these slices on Unix and Windows platforms.

* Support for returning the appropriate `Content-Range` header and HTTP 206 responses for these requests.

* Tests for these features.

It does not support Windows as of yet, largely because I don't have a good test environment -- it should be possible to accomplish by passing some [`OVERLAPPED` structure](https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-overlapped) to `ReadFile()`.

There is also no support for advanced features such as [multipart ranges](https://tools.ietf.org/html/rfc7233#appendix-A) or [the `If-Range` header](https://tools.ietf.org/html/rfc7233#section-3.2).

Since `Range` header support is always optional (a server can just respond with the whole file and a HTTP 200 instead), we just fall back on existing behaviour in these cases instead of issuing some kind of error.

~Along these same lines, we don't yet advertise that we support the `Range` header by sending an [`Accept-Ranges: bytes` header](https://tools.ietf.org/html/rfc7233#section-2.3) on other GET/HEAD requests.~

Closes #259.

atheriel commented:
Conflicts and tests on Windows fixed.

wch commented on src/webapplication.cpp:
```diff
@@ -578,11 +615,22 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
 
   if (method == "HEAD") {
     pDataSource2.reset();
+    hasRange = false;
```
For HEAD requests, it should add the `Accept-Ranges` header, I believe, because that's how clients check if range requests are supported.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests

atheriel replied:
Do we want to advertise range request support even if multipart ranges are not supported?

atheriel replied:
Done.

wch commented on src/webapplication.cpp:
```diff
@@ -563,6 +564,42 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
+      if (start >= pDataSource->size()) {
+        boost::shared_ptr<HttpResponse> pResponse = error_response(pRequest, 416);
+        pResponse->headers().push_back(
+          std::make_pair("Content-Range", std::string("bytes */") + toString(pDataSource->size()))
```
The 416 response uses `size()` to report the file size, but this makes me a bit uneasy, because that's not exactly what `size()` means anymore (as described above). At the very least, there should be a comment about why it happens to work here; in case there are future changes to the code that causes `size()` to change before getting here.

atheriel replied:
I've added a specific `fileSize()` method and used it here to clear up the ambiguity.

wch commented on src/webapplication.cpp:
```diff
@@ -563,6 +564,42 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
+        // unfulfillable requests, so use HTTP 400 for bad syntax instead.
+        return error_response(pRequest, 400);
+      }
+      if (start >= pDataSource->size()) {
```
What about other invalid values, like if `end` is greater than the file, or if `end<start`?

wch replied:
Also, it's good that this check is for `>=`. There should also be a test for cases where the `start` is the same as the size of the file -- which should be an error because the counting starts from 0.

atheriel replied:
Added a test for this :+1:.

wch commented on src/webapplication.cpp:
```diff
@@ -563,6 +564,42 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
+  // Check the HTTP Range header, if it has been specified.
+  std::smatch rangeMatch;
+  int start = 0, end = pDataSource->size() - 1;
+  if (pRequest->hasHeader("Range")) {
```
What happens in the case of a multipart range request?

atheriel replied:
It's ignored and we fall back on the usual `200` behaviour. Since `Range` support is optional, clients should always be equipped to have their request ignored.

I can add a unit test making this explicit.

wch commented on tests/testthat/test-static-paths.R:
```diff
@@ -808,3 +808,69 @@ test_that("Paths with non-ASCII characters", {
+    sprintf("bytes 1000-%d/%d", file_size - 1, file_size)
+  )
+  expect_equal(length(r$content), 303)
+})
```
It would also be good to have tests of edge cases, like requests of length 0 and 1, as well as requests of the offsets 0, 1, the last byte, and the last byte + 1.

atheriel replied:
It's not possible to ask for a 0-size range, but I've added tests for 1-byte reads from the start and end of the files (plus the existing ones that fail for last byte + 1).

wch commented on tests/testthat/test-static-paths.R:
```diff
@@ -808,3 +808,69 @@ test_that("Paths with non-ASCII characters", {
+    parse_headers_list(r$headers)$`content-range`,
+    sprintf("bytes 0-499/%d", file_size)
+  )
+  expect_equal(length(r$content), 500)
```
These tests should also check that the correct content is returned.

atheriel replied:
Done.

wch commented on src/webapplication.cpp:
```diff
@@ -578,11 +615,22 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
   if (client_cache_is_valid) {
     pDataSource2.reset();
     status_code = 304;
+    hasRange = false;
```
I'd prefer not setting this variable here (and also in the previous `if` statement). If possible I'd like to avoid heading down a path where all sorts of logic gets intertwined, because that will make it harder to maintain and debug in the future. It may be reasonably manageable now, but if we want to handle more special logic, it could get difficult.

atheriel replied:
I cleaned up the logic here so that we don't need to do this anymore.

atheriel commented on src/filedatasource-unix.cpp:
```diff
@@ -52,6 +52,18 @@ uint64_t FileDataSource::size() const {
+  if (end > _fsize) {
+    return false;
+  }
+  if (lseek(_fd, start, SEEK_SET) < 0) {
```
From what I can glean from `lseek(2)`, the only error condition applicable here is when `fd` is not an open file descriptor. Otherwise, I'm not sure.

atheriel commented on src/filedatasource.h:
```diff
@@ -20,6 +20,7 @@ class FileDataSource : public DataSource {
 #else
   int _fd;
   off_t _length;
```
Perhaps `_payloadSize`?

atheriel commented on src/filedatasource-win.cpp:
```diff
@@ -63,6 +63,10 @@ uint64_t FileDataSource::size() const {
 }
 
+bool FileDataSource::setRange(size_t start, size_t end) {
+  return false;
```
I do not have a Windows test environment at the moment -- is that a dealbreaker?

wch replied:
Yes -- no Windows support for this would be a dealbreaker. Whenever possible, we don't want to have features that work on one platform and not another. We want code to work consistently across platforms -- it's a bad experience if someone's code works on one computer, but not another. For example, imagine if static file serving worked on Unixes but not Windows. Then someone would complain about it, and then someone is going to be on the hook to implement it.

If there's some technical reason that a feature can't work on Windows, that's one thing. For example, the Windows process model doesn't use `fork`, so some parallel processing code won't work. That's a legitimate reason for only supporting Unix -- but even so, it's an annoying one for users to have to work around (speaking from experience). If there's not a technical reason like that, then the code really should be made to work on all platforms.

atheriel replied:
That's a fair assessment, but it will take me some time. I'll try to resolve the other issues with this PR in the meantime.

atheriel commented on src/filedatasource-unix.cpp:
```diff
@@ -52,6 +52,18 @@ uint64_t FileDataSource::size() const {
   return _length;
 }
 
+bool FileDataSource::setRange(size_t start, size_t end) {
```
Fixed both. The caller verified `end >= start` before, but now it's also checked here.

atheriel commented on src/filedatasource-unix.cpp:
```diff
@@ -35,7 +35,7 @@ FileDataSourceResult FileDataSource::initialize(const std::string& path, bool ow
     }
 
-    _length = info.st_size;
+    _length = _fsize = info.st_size;
```
Fixed.

atheriel commented on src/filedatasource.h:
```diff
@@ -32,6 +33,7 @@ class FileDataSource : public DataSource {
 
   FileDataSourceResult initialize(const std::string& path, bool owned);
   uint64_t size() const;
```
I've added a new `fileSize()` method and documented both to clear up this ambiguity.

atheriel commented on src/webapplication.cpp:
```diff
@@ -563,6 +564,42 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
 
+  // Check the HTTP Range header, if it has been specified.
+  std::smatch rangeMatch;
+  int start = 0, end = pDataSource->size() - 1;
```
Fixed.

atheriel commented on src/webapplication.cpp:
```diff
@@ -578,11 +615,22 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
+    hasRange = false;
+  }
+
+  std::ostringstream contentRange;
```
Fixed.

atheriel commented on src/webapplication.cpp:
```diff
@@ -1,6 +1,7 @@
 #include <boost/bind.hpp>
 #include <boost/shared_ptr.hpp>
 #include <boost/make_shared.hpp>
+#include <regex>
```
I will attempt to convert it to something simpler using `find()`.

atheriel replied:
I've removed `<regex>` use in favour of more hand-rolled parsing. Since `boost::regex` isn't bundled with the `BH` package I thought this was easier, if more verbose.

atheriel commented on src/webapplication.cpp:
```diff
@@ -563,6 +564,42 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
+    // This is a limited form of the HTTP Range header, so don't complain to the
+    // client if it's malformed -- just silently ignore it instead.
+    if (std::regex_search(rangeHeader, rangeMatch, std::regex("bytes=(\\d+)-(\\d+)?$"))) {
+      start = std::stoi(rangeMatch[1].str().c_str());
```
I've updated parsing to use C's `strtoull()` (which doesn't throw) and added handling (and many tests) for malformed input.

atheriel committed:
Add partial support for the 'Range' header when serving static files.

This includes

* Parsing and validation of the most common forms[0] of the `Range`
  header, namely `bytes=0-1000` and `bytes=1001-`, for serving slices of
  files and enabling resumable downloads, etc.

* Support for returning partial files matching these slices on Unix
  and Windows platforms.

* Support for returning the appropriate `Content-Range` header and HTTP
  206 responses for these requests.

* Support for sending the `Accept-Ranges: bytes` header when appropriate.

* Tests for these features.

There is no support for suffix length ranges, multipart ranges[2] or the
`If-Range` header[3].

Since `Range` header support is always optional (a server can just
respond with the whole file and a HTTP 200 instead), we just fall back
on existing behaviour in these cases instead of issuing some kind of
error.

Closes #259.

[0]: https://tools.ietf.org/html/rfc7233#section-3.1
[2]: https://tools.ietf.org/html/rfc7233#appendix-A
[3]: https://tools.ietf.org/html/rfc7233#section-3.2
[4]: https://tools.ietf.org/html/rfc7233#section-2.3

atheriel commented on src/webapplication.cpp:
```diff
@@ -563,6 +564,42 @@ boost::shared_ptr<HttpResponse> RWebApplication::staticFileResponse(
     }
   }
 
+  // Check the HTTP Range header, if it has been specified.
```
It's a bit awkward to do so, since there are currently three variables set in this section.

wch replied:
It would be simpler to review and maintain the code if the range header parsing was a separate function. Right now the string parsing is mixed in with the logic for setting range and generating error responses; if the string parsing were done in a separate step, that would make the code more manageable. Even if the code is correct now, I would feel uneasy about touching it in the future.

Something like this would be easier to review and maintain:

```c++
std::pair<int64_t, int64_t> parseRangeHeader(std::string rangeHeader) {
  if (rangeHeader.find("bytes=") != 0) {
    return std::make_pair<int64_t, int64_t>(-1, -1);
  }

  rangeHeader = rangeHeader.substr(6);
  int sep_offset = rangeHeader.find("-");
  if (sep_offset == std::string::npos) {
    return std::make_pair<int64_t, int64_t>(-1, -1);
  }

  std::string start_str = rangeHeader.substr(0, sep_offset+1);
  std::string end_str   = rangeHeader.substr(sep_offset+1);
  
  int64_t start;
  int64_t end;

  // TODO: Do number parsing here; return (-1, -1) if either start or end are not
  // well-formed.
  
  return std::make_pair<int64_t, int64_t>(start, end);
}
```

The number parsing could be done with `strtoull()` as you currently have, or with `sscanf()`. (There are a number of details to get right with those functions and I don't have the time right now to try to do it, so I didn't implement that part.) It might make sense to extract it out into a function.

sounkou-bioinfo commented:
Hi all,
What is the status of this PR, this would be extremely useful in contexts like serving parquet files or other files for which partial downloads are supported. Use case do work on the client side instead of server for sat parquet files or serving duckdb databases
